### ISSUE TITLE
[Idea] Adding Kubevirt Support for Local Development

### ISSUE THREAD

**Issue by afritzler** [gardener/gardener#827]
Hi all,

since the work on the Kubevirt support for Gardener is progressed quite nicely[0], I wanted to suggest the following:

How about adding the Kubvirt Support into Gardener and MCM (currently everything is a hard fork)? The idea here would be to use your local minikube as the Gardener Cluster, the seed + the "IaaS" thanks to Kubevirt. 

In your minikube you would additionally install:
* Kubevirt controller
* Metal-lb (need to check how good that works)
* Using nip.io for DNS 

The advantage of this setup in contrast to the Vagrant based one we currently have is, that from an architectural point of view it fits into the Gardener model like any other IaaS we support. Of course you could still use Kubevirt in a non-local dev setup as well.

I just wanted to start a discussion here and see what you guys think and if there would be need for such thing? Otherwise we can postpone that and do the integration via our extension model.

Regards,
Andreas

/cc @gonzolino 

[0] https://github.com/kubeforge

**Comment by vlerenc** [comment]
That sounds actually like a very elegant solution to the issue of local setup. Just a [Minikube](https://kubernetes.io/docs/setup/minikube) on whatever, [Kind](https://github.com/kubernetes-sigs/kind), or [Docker Desktop for Mac](https://www.docker.com/products/docker-desktop) and off you go as garden and seed cluster and even IaaS? If that would work, it would rock so much! It would be a truly great story to tell people about. So elegant, so cool!

💯👍

**Comment by rfranzke** [comment]
Nice idea, +100 if that really works that smooth. However, I suggest to wait for the extension concepts to be in place before we now add another provider that needs to be factored out later again. OK? :-)

**Comment by afritzler** [comment]
No problem. We need to test a few things here anyways. So we will keep the forks for now and aim for the integration via the extension concept.

**Comment by fabiand** [comment]
Side note, that we're still aiming at kubernetes/minikube#2178 - which should be more in reach these days than back then.

**Comment by rfranzke** [comment]
Interesting, thank's for pointing this out!

**Comment by afritzler** [comment]
Thanks for the link @fabiand. As you pointed out, the kubevirt operator makes it super easy to use kubevirt inside minikube. The only thing I had to use though is, setting the emulation mode to true.

```
apiVersion: v1
kind: ConfigMap
metadata:
  namespace: kubevirt
  name: kubevirt-config
data:
  debug.useEmulation: "true"
```

This works in a Hyperkit or KVM2 box. Since Virtualbox 6 supports nested virtualization, this is probably not necessary there.

**Comment by fabiand** [comment]
Yes - This would actually a part we could eventually do automagically in the add-on code, to enable emulation if the driver != kvm2.

**Comment by pablochacin** [comment]
That would be great. Not having to use Vagrant will simplify testing locally the images for different operating systems.

**Comment by fabiand** [comment]
Now, despite how great this is - My issue is that we'll not get to this
within weeks.

**Comment by mkoderer** [comment]
@afritzler do you need persistent volumes in such a setup?

**Comment by afritzler** [comment]
@mkoderer not from day one. We can start without having an answer for dynamic storage in the kubevirt vms.

**Comment by rfranzke** [comment]
@gonzolino any update on the kubevirt extension for Gardener?

**Comment by gonzolino** [comment]
We started development here some time ago: https://github.com/kubeforge/gardener-extensions/tree/feature.added_kubevirt_extenstion
But there hasn't been any progress for quite some time now 😕

**Comment by rfranzke** [comment]
:( Any plans to continue with it in the next weeks?

**Comment by gonzolino** [comment]
If there is interest in it @afritzler and I can do another joint session to work on it :)

**Comment by afritzler** [comment]
I think the interest is definitely there. We just need to push forward with the kubevirt extension story.

**Comment by rfranzke** [comment]
The interest is very high because it would simplify the local dev. setup. So far, when developing Gardener, you have to deploy to real clusters at "real" providers.
